Feature/metrics api epmdj 10827

iryabov opened the pull request.

RomanDavlyatshin committed:
feat: add sql function to query instances by branch + first instance of base branch

RomanDavlyatshin committed:
feat: instance ids by branch use agent & groupId + simplify query

RomanDavlyatshin committed:
feat: sql fun for packages/classes/methods + risks by branch diff

RomanDavlyatshin committed:
feat: migrations for vcs_metadata, timestamps, metrics

RomanDavlyatshin committed:
feat: same-branch metrics + optional baseline commit hash param

RomanDavlyatshin committed:
build: set sharedLibsRef to feature/add-vcs-metadata-parameters

iryabov committed:
fix: add vcs* parameters to AgentConfigTable

EPMDJ-10811

iryabov committed:
feat: add repeatable migrations for sql metric calculation functions

EPMDJ-10811

iryabov committed:
fix: add schema prefix for `get_instance_ids_by_branch` function calls

EPMDJ-10811

iryabov committed:
feat: add metrics api

EPMDJ-10827

iryabov committed:
feat: add metrics repository

EPMDJ-10827

iryabov committed:
refactor: move metrics api to the module `admin-metrics`

EPMDJ-10828

RomanDavlyatshin commented on admin-metrics/src/main/resources/raw_data/db/migration/R__Metrics_Calculation_Functions.sql:
tbd `R__` prefix

RomanDavlyatshin commented on admin-metrics/src/main/kotlin/com/epam/drill/admin/metrics/config/MetricsDatabaseConfig.kt:
```diff
@@ -0,0 +1,82 @@
+import java.sql.ResultSet
+import javax.sql.DataSource
+
+object MetricsDatabaseConfig {
```
_tbd_ fine tuning connections numbers + best performance config

RomanDavlyatshin commented on admin-metrics/src/main/resources/raw_data/db/migration/R__Metrics_Calculation_Functions.sql:
```diff
@@ -0,0 +1,612 @@
+        UNION ALL
+        SELECT
+            t.id,
+			t.agent_id,
```
odd formatting

RomanDavlyatshin commented on admin-metrics/src/main/resources/raw_data/db/migration/V1__Raw_data_schema_init.sql:
_tbd_ duplicate migrations (we have the exact same files in admin-writer module)

RomanDavlyatshin commented on admin-writer/src/main/kotlin/com/epam/drill/admin/writer/rawdata/service/Coverage.kt:
```diff
@@ -1,1075 +0,0 @@
-    return result
-}
-
-fun generateHtmlTable(results: List<JsonObject>): String {
```
I suggest to keep / reimplement something along these lines for our new routes
To provide simple way to get JSON/HTML reports out of our system

RomanDavlyatshin commented on admin-app/src/main/kotlin/com/epam/drill/admin/DrillAdminApplication.kt:
```diff
@@ -101,6 +105,8 @@ fun Application.module() {
                 }
             }
+
+            metricRoutes()
```
That is definitely should be under`authenticate("api-key")`

RomanDavlyatshin commented on admin-metrics/src/main/kotlin/com/epam/drill/admin/metrics/route/MetricRoutes.kt:
```diff
@@ -0,0 +1,107 @@
+        val currentBranch: String,
+        val currentVcsRef: String,
+        val baseBranch: String,
+        val baseVcsRef: String,
```
`baseVcsRef` must be optional

RomanDavlyatshin committed:
chore: delete unnecessary sql migration files from admin-metrics
WHY: they were added by mistake

RomanDavlyatshin committed:
feature: allow coverage metrics calculation by instance_id
Now metrics can be calculated for either agent-id+group-id+commit or just instance_id

RomanDavlyatshin committed:
chore: rename sql migration file to adhere to migrations naming schema

RomanDavlyatshin committed:
feat: WIP /builds,/instances,/methods,/coverage API
TODO: move payloads to separate package
TODO: method id from buildId + signature
TODO: deserialize probes to Array<Boolean>

RomanDavlyatshin committed:
chore: remove commented-out lib-jvm-shared dependencies

RomanDavlyatshin committed:
feat: extract api payloads to own package

RomanDavlyatshin committed:
feat: sql - rename exec_class_data and ast_method to coverage and methods respectively

RomanDavlyatshin committed:
fix: instance.id to instance.buildId typo

RomanDavlyatshin committed:
feat: make admin-writer payloads and entities classes instead of data classes

RomanDavlyatshin committed:
chore: add license header

RomanDavlyatshin committed:
refactor: fix interface param name mismatch warning

RomanDavlyatshin committed:
chore: add "out" folders to .gitignore

RomanDavlyatshin committed:
fix: sql typos + remove old metrics calc function (to be refactored)

RomanDavlyatshin committed:
feat: WIP deserialize and store boolean array probes

RomanDavlyatshin committed:
feat: move method id generation to service + fix missing build_id in MethodTable

iryabov committed:
fix: change Array of Boolean to BooleanArray in the coverage API and the coverage repository

iryabov committed:
build: remove lib-jvm-shared scripts and files

RomanDavlyatshin committed:
deps: update to kotlin 2.0.0, exposed 0.50.1
TODO: replace usage of deprecated select  in DatabaseUserRepository and DatabaseApiKeyRepository

RomanDavlyatshin committed:
fix: replace exposed's deprecated .select{} with .selectAll().where

iryabov committed:
refactor: upgrade Ktor up to 2.2.4
refactor: upgrade Kotlin up to 1.8.22 and other Kotlin libs

EPMDJ-10838

RomanDavlyatshin committed:
build: set logback:1.3.14

RomanDavlyatshin committed:
feat: json serializer - treat missing nullable fields as null values

RomanDavlyatshin committed:
feat: json serializer - treat missing nullable fields as null values

RomanDavlyatshin committed:
build: remove unnecessary gradle build options

RomanDavlyatshin committed:
fix: upsert for /builds, /instances, /methods + remove commitTags field + save commitMessage properly

RomanDavlyatshin committed:
docs: remove mentions of lib-jvm-shared
TODO: edit release GH action

RomanDavlyatshin committed:
build: gradle.properties - remove unused values

RomanDavlyatshin committed:
fix: make commitSha, buildVersion, instanceId fields nullable - InstancePayload, MethodPayload

RomanDavlyatshin committed:
feat: insert build data on PUT /instances if corresponding build's data is missing

RomanDavlyatshin committed:
chore: add .kotlin/ to .gitignore

iryabov committed:
Merge remote-tracking branch 'origin/feature/metrics-api-EPMDJ-10827' into feature/ktor-upgrade-EPMDJ-10838

# Conflicts:
#	admin-auth/src/main/kotlin/com/epam/drill/admin/auth/config/JwtConfig.kt
#	admin-auth/src/main/kotlin/com/epam/drill/admin/auth/repository/impl/DatabaseApiKeyRepository.kt
#	admin-writer/src/main/kotlin/com/epam/drill/admin/writer/rawdata/route/RawDataWriterRoutes.kt
#	gradle.properties

iryabov committed:
chore: remove OptInt annotation as it is no longer needed

EPMDJ-10838

iryabov committed:
refactor: upgrade Ktor up to 2.3.11
refactor: add explicitNulls = false in Json configuration

EPMDJ-10838

RomanDavlyatshin committed:
Merge pull request #386 from Drill4J/feature/ktor-upgrade-EPMDJ-10838

Feature/ktor upgrade epmdj 10838

RomanDavlyatshin committed:
feature: created_at for instances, builds, coverage, tests

iryabov committed:
fix: use `decompressAndReceive` function in routes to explicitly decompress the request body

iryabov committed:
refactor: replace deprecated `@Location` to  `@Resource`

EPMDJ-10838

iryabov committed:
refactor: remove unused deprecated class `Login`

EPMDJ-10838

iryabov committed:
refactor: change deprecated function `appendExpired` to actual `append`

EPMDJ-10838

iryabov committed:
refactor: move to a new ktor testing api in admin-app module

EPMDJ-10838

RomanDavlyatshin committed:
feature: adjust raw_data.tests SQL & classes for new API

RomanDavlyatshin committed:
feature: coverage report SQL functions

iryabov committed:
refactor: improve exception message when content type is missing

EPMDJ-10838

iryabov committed:
Merge pull request #388 from Drill4J/fix/return-gzip

fix: use `decompressAndReceive` function in routes to explicitly deco…

iryabov committed:
Merge remote-tracking branch 'origin/feature/metrics-api-EPMDJ-10827' into feature/replace-locations-EPMDJ-10838

# Conflicts:
#	admin-writer/src/main/kotlin/com/epam/drill/admin/writer/rawdata/route/RawDataWriterRoutes.kt

iryabov committed:
fix: set a cookie value to empty string instead "/"

EPMDJ-10838

iryabov committed:
Merge pull request #389 from Drill4J/feature/replace-locations-EPMDJ-10838

Feature/replace locations epmdj 10838

RomanDavlyatshin committed:
feat: add data-ingest prefix for raw data api

RomanDavlyatshin committed:
feat: build risks accumulated coverage SQL function

RomanDavlyatshin committed:
feat: add method signature field

RomanDavlyatshin committed:
feat: add function get_recommended_tests

RomanDavlyatshin commented:
already in #391
